render/src/filters.rs

```diff
@@ -66,3 +167,288 @@ impl Default for ColorMatrixFilter {
         }
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct ConvolutionFilter {
+    pub bias: f32,
+    pub clamp: bool,
+    pub default_color: Color,
+    pub divisor: f32,
+    pub matrix: Vec<f32>,
+    pub matrix_x: u8,
+    pub matrix_y: u8,
+    pub preserve_alpha: bool,
+}
+
+impl From<&swf::ConvolutionFilter> for ConvolutionFilter {
+    fn from(value: &swf::ConvolutionFilter) -> Self {
+        let preserve_alpha = value.is_preserve_alpha();
+        let clamp = value.is_clamped();
+        Self {
+            bias: value.bias.to_f32(),
+            clamp,
+            default_color: value.default_color.clone(),
+            divisor: value.divisor.to_f32(),
+            matrix: value.matrix.iter().map(|v| v.to_f32()).collect(),
+            matrix_x: value.num_matrix_cols,
+            matrix_y: value.num_matrix_rows,
+            preserve_alpha,
+        }
+    }
+}
+
+impl Default for ConvolutionFilter {
+    fn default() -> Self {
+        Self {
+            bias: 0.0,
+            clamp: true,
+            default_color: Color::from_rgba(0),
+            divisor: 1.0,
+            matrix: vec![],
+            matrix_x: 0,
+            matrix_y: 0,
+            preserve_alpha: true,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum DisplacementMapFilterComponent {
+    Alpha,
+    Blue,
+    Green,
+    Red,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum DisplacementMapFilterMode {
+    Clamp,
+    Color,
+    Ignore,
+    Wrap,
+}
+
+#[derive(Debug, Clone)]
+pub struct DisplacementMapFilter {
+    pub color: Color,
+    pub component_x: u8,
+    pub component_y: u8,
+    pub map_bitmap: Option<BitmapHandle>,
+    pub map_point: (i32, i32),
+    pub mode: DisplacementMapFilterMode,
+    pub scale_x: f32,
+    pub scale_y: f32,
+}
+
+impl Default for DisplacementMapFilter {
+    fn default() -> Self {
+        Self {
+            color: Color::from_rgba(0),
+            component_x: 0,
+            component_y: 0,
+            map_bitmap: None,
+            map_point: (0, 0),
+            mode: DisplacementMapFilterMode::Wrap,
+            scale_x: 0.0,
+            scale_y: 0.0,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DropShadowFilter {
```

Comment:
I wonder if we could use `swf::DropShadowFilter` etc. directly without an intermediate struct.

Reply:
That was what I used originally but there's not a 1:1 between swf filter tags and actual flash supported filters.

Reply:
All filter types in `render/src/filters.rs` in this PR has a SWF counterpart. Am I missing something?

Reply:
ShaderFilter (currently unimplemented) and DisplacementMapFilter do not exist in swf form.

There's also states that exist in swf that are somewhat nonsense to the renderer, like bevel filter that's both inside and on top, or convolution filter with a Vec that doesn't match rows*cols.

Technically the swf structs are lossy vs actionscript made filters, with Fixed16s, but I don't know if that'll actually matter in principal though.

Reply:
I've pushed a commit that does that. I've had to lower the accuracy of the tests significantly to account for the usage of Fixed8/Fixed16 instead of f32 now though.